tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,553 @@
+          auto ios = GetInputusAndOutputsOfOp(op);
+          for (auto v : ios) {
+            Value ref = ios[0];
+            if (shape_checker_[ref].Get() != shape_checker_[v].Get()) {
+              converged = false;
+              shape_checker_[ref].Merge(&shape_checker_[v]);
+            }
+          }
+        }
+      }
+    } while (!converged);
+  }
+
+  SmallVector<Operation*, 4> op_list_;
+  DenseMap<Value, UnionFind<Value>> shape_checker_;
+};
+
+class FusionPlanner {
+ public:
+  FusionPlanner(const SmallVector<Operation*, 4>& op_list)
+   : op_list_(op_list), shape_analysis_(op_list) {
+    BuildNodeMap();
+  }
+
+  llvm::Optional<FusionPlan> Run() {
+    // Greedily search kLoop/kInput pattern
+    RunEdgeContractionLoop();
+
+    FusionPlan plan;
+    DenseMap<int, int> pattern_ids;
+    for (auto op : op_list_) {
+      Cluster* cluster = GetClusterForNode(op);
+      auto node_id = cluster->cycles_graph_node_id();
+      if (!IsFusible(op_list_[node_id])) {
+        continue;
+      }
+      if (!pattern_ids.count(node_id)) {
+        int pattern_id = pattern_ids.size();
+        pattern_ids[node_id] = pattern_id;
+        plan.emplace_back();
+      }
+      plan[pattern_ids[node_id]].push_back(op);
+    }
+    return plan;
+  }
+
+  const SmallVector<Operation*, 4>& op_list() const {
+    return op_list_;
+  }
+
+ private:
+  class Cluster {
+   public:
+    Cluster(int node_id, FusionPlanner* planner)
+        : node_id_(node_id), planner_(planner) {
+      auto& op_list = planner->op_list();
+      pattern_.push_back(op_list[node_id]);
+    }
+
+    // Merges `other` into this cluster, and clears `other`.
+    void Merge(Cluster* other) {
+      pattern_.insert(pattern_.end(),
+          other->pattern_.begin(), other->pattern_.end());
+    }
+
+    // The number of nodes in this cluster.
+    int cluster_size() const { return pattern_.size(); }
+
+    // The ID of the cluster as represented in `cycles_graph_`.
+    int cycles_graph_node_id() const { return node_id_; }
+
+    FusionPattern fused_pattern() {
+      return pattern_;
+    }
+
+   private:
+    int node_id_;
+    FusionPlanner* planner_;
+
+    FusionPattern pattern_;
+  };
+
+ private:
+  Cluster* MakeCluster(int cycles_graph_node_id) {
+    cluster_storage_.emplace_back(
+        new Cluster(cycles_graph_node_id, this));
+    return cluster_storage_.back().get();
+  }
+
+  void BuildNodeMap() {
+    for (auto& op : op_list_) {
+      auto node_id = cycle_detector_.NewNode();
+      op_to_node_id_[op] = node_id;
+      cluster_for_node_.emplace_back();
+      cluster_for_node_[node_id].Get() = MakeCluster(node_id);
+      for (int idx = 0; idx < op->getNumOperands() ; ++idx) {
+        auto operand_op = op->getOperand(idx).getDefiningOp();
+        if (operand_op == nullptr) {
+          // skip block argument
+          continue;
+        }
+        auto iter = op_to_node_id_.find(operand_op);
+        assert(iter != op_to_node_id_.end());
```

Comment:
Be cautious with using assert

Reply:
Do you believe that this assert can be hit by any IR input to the pass?